sklearn/compose/_column_transformer.py

```diff
@@ -411,33 +409,17 @@ def _validate_output(self, result):
                     "The output of the '{0}' transformer should be 2D (scipy "
                     "matrix, array, or pandas DataFrame).".format(name))
 
-    def _validate_features(self, n_features, feature_names):
+    def _validate_features(self, feature_names):
         """Ensures feature counts and names are the same during fit and
         transform.
-
-        TODO: It should raise an error from v0.24
         """
 
-        if ((self._feature_names_in is None or feature_names is None)
-                and self._n_features == n_features):
+        if self._feature_names_in is None or feature_names is None:
             return
 
-        neg_col_present = np.any([_is_negative_indexing(col)
-                                  for col in self._columns])
-        if neg_col_present and self._n_features != n_features:
-            raise RuntimeError("At least one negative column was used to "
-                               "indicate columns, and the new data's number "
-                               "of columns does not match the data given "
-                               "during fit. "
-                               "Please make sure the data during fit and "
-                               "transform have the same number of columns.")
-
-        if (self._n_features != n_features or
-                np.any(self._feature_names_in != np.asarray(feature_names))):
-            warnings.warn("Given feature/column names or counts do not match "
-                          "the ones for the data given during fit. This will "
-                          "fail from v0.24.",
-                          FutureWarning)
+        if np.any(self._feature_names_in != np.asarray(feature_names)):
```

Comment:
`np.asarray(feature_names)` would create a copy. To prevent the copy we can:

```py
if any(left != right
        for left, right in zip(self._feature_names_in, feature_names)):
```

Reply:
Actually ``_validate_features`` is only called in transform where the feature names have already been through ``np.asarray`` so ``if np.any(self._feature_names_in != feature_names):`` should be fine.

Reply:
I was going to make the same remark. Actually now it makes more sense to have the validation directly in `transform` because it is a single `if` statement and a `raise`. It will be more obvious that we deal with ndarray because it is next to the conversion.